package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs

```diff
@@ -486,44 +492,67 @@ static void FlattenPrimvar(Primvar<object> primvar)
-                    if (changeHandedness)
+                newFacesCount += faceVertexCounts[i] - 2;
+            }
+
+            var newValues = new T[3 * newFacesCount]; // we're converting every face to a triangle, so must be three verts per face
+
+            int last = 0, currentIndex = 0;
+            if (changeHandedness)
+            {
+                for (var faceIndex = 0; faceIndex < faceVertexCounts.Length; faceIndex++)
+                {
+                    var next = last + 1;
+                    for (var triangleIndex = 0; triangleIndex < faceVertexCounts[faceIndex] - 2; triangleIndex++)
                     {
-                        newValues.Add(values[next++]);
-                        newValues.Add(values[last]);
-                        newValues.Add(values[next]);
+                        newValues[currentIndex++] = values[next++];
+                        newValues[currentIndex++] = values[last];
+                        newValues[currentIndex++] = values[next];
                     }
-                    else
+                    last += faceVertexCounts[faceIndex];
+                }
+            }
+            else
+            {
+                for (var faceIndex = 0; faceIndex < faceVertexCounts.Length; faceIndex++)
+                {
+                    var next = last + 1;
+                    for (var triangleIndex = 0; triangleIndex < faceVertexCounts[faceIndex] - 2; triangleIndex++)
                     {
-                        newValues.Add(values[last]);
-                        newValues.Add(values[next++]);
-                        newValues.Add(values[next]);
+                        newValues[currentIndex++] = values[last];
+                        newValues[currentIndex++] = values[next++];
+                        newValues[currentIndex++] = values[next];
                     }
-
-                last += faceVertexCount[i];
+                    last += faceVertexCounts[faceIndex];
+                }
             }
 
-            values = newValues.ToArray();
+            values = newValues;
         }
 
         /// <summary>
         /// Utility method to convert a given array of values to the equivalent faceVarying array (one value per vertex per face).
         /// </summary>
-        /// <remarks> If the interpolation of the array to convert is not known, it will be guessed based on the length of the array. </remarks
+        /// <remarks> If the interpolation of the array to convert is not known, it will be guessed based on the length of the array. </remarks>
         void ConvertInterpolationToFaceVarying<T>(ref T[] values, int[] vertexIndices, bool changeHandedness = false, TfToken interpolation = null)
         {
             if (values == null)
                 return;
 
-            if (interpolation == null)
-                interpolation = GuessInterpolation(values.Length);
+            var newInterpolation = GuessInterpolation(values.Length);
```

Comment:
it seems like we now call GuessInterpolation no matter what the given interpolation was - i am guessing that when interpolation is given it should in most cases match the result of GuessInterpolation? 

Is this done as validation?
if so, is GuessInterpolation a quick process? If it is, do we now need a "Skip Interpolation Validation" flag for more experienced users?

Reply:
Yep, that's the intention. It turns out now that we have arbitrary primvars we can't trust the authoring DCC of the USD file to actually have set the correct interpolation token. In the majority of cases no interpolation is passed in, so it was always performing the 'guess' anyway. GuessInterpolation is pretty cheap and this won't be a particularly hot path so I don't think we need to expose it to users. This whole area is quite hacky because we're forcing all the attributes to be face varying instead of just using what they're authored as.